File: Scala/algorithmic_toolbox/src/main/scala-2.12/dynamicprogramming/PrimitiveCalculator.scala
package dynamicprogramming

import scala.annotation.tailrec

object PrimitiveCalculator {
  private val Nominators: List[Int] = List(2, 3)

  private def findPreviousMinimum(minOperations: Array[Int], k: Int): (Int, Int) = {
    def checkForShorterCalculation(acc: (Int, Int), nominator: Int): (Int, Int) = {
      val (previousMinimum, _): (Int, Int) = acc
      if (k % nominator == 0) {
        val pos: Int = k / nominator - 1
        val nrOps: Int = minOperations(pos)
        if (nrOps < previousMinimum) (nrOps, pos)
        else acc
      }
      else acc
    }

    val position: Int = k - 2
    Nominators.foldLeft((minOperations(position), position))(checkForShorterCalculation)
  }

  private def backtrackCalculation(backtrack: Array[Int], n: Int): List[Int] = {
    @tailrec
    def loop(k: Int, path: List[Int]): List[Int] = {
      if (k <= 1) path
      else {
        val m: Int = backtrack(k)
        loop(m, (m + 1) :: path)
      }
    }
    loop(n - 1, List(n))
  }

  def runCalculator(n: Int): List[Int] = {
    val minOperations: Array[Int] = Array.fill(n)(0)
    val backtrack: Array[Int] = Array.fill(n)(0)
    for { k <- 2 to n } {
      val (previousMinimum, position): (Int, Int) = findPreviousMinimum(minOperations, k)
      minOperations(k - 1) = previousMinimum + 1
      backtrack(k - 1) = position
    }
    backtrackCalculation(backtrack, n)
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val n: Int = reader.next().toInt
    val result: List[Int] = runCalculator(n)
    println(result.length - 1)
    println(result.mkString(" "))
  }
}

File: OCaml/algorithmic_toolbox/dynamicprogramming/primitive_calculator.ml
let nominators = [2; 3]


let backtrack_calculation backtrack n =
    let rec loop k path =
        if k <= 1 then path
        else
            let m = backtrack.(k) in
            loop m (m + 1 :: path)
    in loop (n - 1) [n]


let find_previous_minimum min_operations k =
    let check_for_shorter_calculation acc nominator =
        let previous_minimum, _ = acc in
        if k mod nominator = 0 then
            let pos = k / nominator - 1 in
            let nr_ops = min_operations.(pos) in
            if nr_ops < previous_minimum then (nr_ops, pos)
            else acc
        else acc in
    let position = k - 2
    in List.fold_left check_for_shorter_calculation ((min_operations.(position), position))  nominators


let run_calculator n =
    let min_operations = Array.make n 0 in
    let backtrack = Array.make n 0 in
    for k=2 to n do
        let previous_minimum, position = find_previous_minimum min_operations k in
        min_operations.(k - 1) <- previous_minimum + 1;
        backtrack.(k - 1) <- position
    done;
    backtrack_calculation backtrack n


let () =
    let n = read_int() in
    let result = run_calculator n in
    print_int (List.length result - 1);
    print_newline();
    List.iter (fun elem -> print_int elem; print_string " ") result;
    print_newline()
